model/simple/dowhilebehavior.go

```diff
@@ -0,0 +1,143 @@
+	logger := ctx.FlowLogger()
+
+	if ctx.Status() == model.TaskStatusSkipped {
+		return model.EvalDone, nil //todo introduce EVAL_SKIP?
+	}
+
+	if logger.DebugEnabled() {
+		logger.Debugf("Eval do-while Task '%s'", ctx.Task().ID())
+	}
+
+	done, err := ctx.EvalActivity()
+
+	if err != nil {
+		// check if error returned is retriable
+		if errVal, ok := err.(*activity.Error); ok && errVal.Retriable() {
+			// check if task is configured to retry on error
+			retryData, rerr := GetRetryData(ctx, RetryOnErrorAttr)
+			if rerr != nil {
+				return model.EvalFail, rerr
+			}
+			if retryData.Count > 0 {
+				return DoRetry(ctx, retryData, RetryOnErrorAttr), nil
+			}
+		}
+		ref := activity.GetRef(ctx.Task().ActivityConfig().Activity)
+		ctx.FlowLogger().Errorf("Error evaluating activity '%s'[%s] - %s", ctx.Task().ID(), ref, err.Error())
+		ctx.SetStatus(model.TaskStatusFailed)
+		return model.EvalFail, err
+	}
+
+	if !done {
+		ctx.SetStatus(model.TaskStatusWaiting)
+		return model.EvalWait, nil
+	}
+	return dw.checkDoWhileCondition(ctx)
+}
+
+// PostEval implements model.TaskBehavior.PostEval
+func (dw *DoWhileTaskBehavior) PostEval(ctx model.TaskContext) (evalResult model.EvalResult, err error) {
+	ctx.FlowLogger().Debugf("PostEval do-while Task '%s'", ctx.Task().ID())
+
+	_, err = ctx.PostEvalActivity()
+
+	if err != nil {
+		// check if error returned is retriable
+		if errVal, ok := err.(*activity.Error); ok && errVal.Retriable() {
+			// check if task is configured to retry on error
+			retryData, rerr := GetRetryData(ctx, RetryOnErrorAttr)
+			if rerr != nil {
+				return model.EvalFail, rerr
+			}
+			if retryData.Count > 0 {
+				return DoRetry(ctx, retryData, RetryOnErrorAttr), nil
+			}
+		}
+		ref := activity.GetRef(ctx.Task().ActivityConfig().Activity)
+		ctx.FlowLogger().Errorf("Error post evaluating activity '%s'[%s] - %s", ctx.Task().ID(), ref, err.Error())
+		ctx.SetStatus(model.TaskStatusFailed)
+		return model.EvalFail, err
+	}
+	return dw.checkDoWhileCondition(ctx)
+}
+
+func (dw *DoWhileTaskBehavior) checkDoWhileCondition(ctx model.TaskContext) (evalResult model.EvalResult, err error) {
+	// check if task is configured to repeat on condition match
+	condition, err := dw.getDoWhileCondition(ctx)
+	if err != nil {
+		return model.EvalFail, err
+	}
+	if len(condition) > 0 {
+		return dw.evaluateCondition(ctx, condition)
+	}
+	return model.EvalDone, nil
+}
+
+// Returns condition set for do-while task type
+func (dw *DoWhileTaskBehavior) getDoWhileCondition(ctx model.TaskContext) (string, error) {
+	if _, ok := ctx.GetWorkingData(DoWhileAttr); !ok {
+		// first attempt - condition attribute not set yet
+		if d, ok := ctx.GetSetting("do-while"); ok {
+			doWhileObj, err := coerce.ToObject(d)
+			if err != nil {
+				return "", fmt.Errorf("do-while setting is not properly configured")
+			}
+			condition, _ := doWhileObj["condition"].(string)
+			ctx.SetWorkingData(DoWhileAttr, condition)
+			return condition, nil
```

Comment:
@cpramodroy I don't think we need to save condition to working data. condition itself from task setting and it should never be updated when within dowhile lopp

Reply:
I was saving it to working data to avoid going through the model every time to get it. If going through the model in each iteration is preferred, I can change it to that.